### ISSUE TITLE
Expanding/collapsing nodes is not announced for JAB objects

### ISSUE THREAD

**Issue by rpaquay** [nvaccess/nvda#5812]
When using a Java application using JAB for accessibility (for example SwingSet2, IntelliJ or Android Studio), nvda does not announce anything when a node in a tree view is collapsed/expanded.

This is related to bug #5766, but the fix for it (#5810) did not include a fix for this specific issue, because it is non-trivial.

I believe the reason this issue is specific to nvda is because:

* I can see events when nodes are expanded/collapsed with the Access Bridge Explorer tool (https://github.com/google/access-bridge-explorer), by enabling the "Options | Capture Events | PropertyStateChange"

* I looked at the nvda code, and I think the issue is related to the fact tree view item nodes are `transient`, because Tree View have the `manage descendants` state. What this means is that when the `PropertyStateChange` event is raised, the source node passed to the event handler (the tree view item) is not the same `JABContext` object as the currently focused element from `API.getFocusedObject`. So, the default `event_statechange` event handler does nothing when processing the event (see https://github.com/nvaccess/nvda/blob/284e670719b0a3e054bceef8a696df3c741ed5b3/source/NVDAObjects/__init__.py#L845)

I am not sure how to handle this issue, as it seems it could be quite difficult to reconcile both sides: `api.getFocusedObject` depends on object identity while JAB tree view items are transient.

**Comment by jcsteh** [comment]
Ug. This is total insanity. It's ridiculous to not provide a proper way for an AT to compare objects, at least with the focus. Sure, an event gets fired, but that's not enough because a screen reader shouldn't just announce all changes.

@mohammad-suliman provided a potential solution for this (comparing the context info) in #5815. I have some concerns about that solution, but my big question is: if these objects really are transient, how are the states the same when comparing the old object with the new? They have to be the same in order for this solution to work, but if the old object is dead, the states shouldn't update and os the comparison should fail. So if the old object isn't dead, that means it isn't really transient after all...?

**Comment by rpaquay** [comment]
@jcsteh My understanding of "transient" is the following -- apologies in advance, this is going to be a long explanation:

* First of all this notion of "transient" is only for children of accessible contexts (i.e. JABContext) that have the "manage descendants" state. Looking at examples in Swing, I see this state only for lists, trees and tables. What this means is that it is ok to use "isSameObject" for accessible contexts which are not transient.

* The accessible context returned for a "transient" child acts as a "proxy" to the actual child element, meaning that when screen readers asks for any information for such a child via its accessible context "proxy", the proxy re-acquires the actual child element (think the Swing element), asks that element about the relevant information and return that information. This is why it is always ok to get information for a transient object: JAB ensures the underlying component if re-aquired/re-instantiated if needed. The accessible context "proxy" is itself not transient, it is the underlying Swing component that is transient.

* That said, given a transient underlying child, JAB creates a new accessible context "proxy' each time the child needs to be exposed via the JAB. That is why when a "stateChange" or "focusGained" event is fired, the "source" parameter is a new accessible context proxyinstance. That explains why the "isSameObject" call fails. But event though it is a distinct instance, it will return the same information when the screen readers asks for location, name, etc., as per described in the bullet above on how "proxy" work.

* Will all that said, it is not clear to me how a screen reader is supposed to compare two accessible context "proxy" instances for equality.
What was done in #5815 seems reasonable, although I would probably add more things to compare, for example I would ensure both elements have the same object depth and that they have the same parent (or grand-parent, etc. for treeviews) and that both parents are both marked as "manage descendants". It is somewhat annoying that while JAB marks the parent as "manage descendants" to notify all children will be transient, there does not seem to be any way to ask directly a child if it is "transient". It seems the only way is to find if any parent is marked as "manage descendants" (yuk!).
A way I can think of to decrease the amount of JAB calls done when a `stateChange` event is received is to check whether the `source` component has the `focused` state. If it does not, we can obviously ignore the event right away without further ado.

* Now, in addition to all of that, I believe JAB has a bug with JTables. From what I can gather, the behavior of the accessible context "proxy" instances for table cells don't correctly implement the "re-acquire" behavior described above. I have been observing this with Access Bridge Explorer. If you expand a "table" node, you get a list of children that represent all the cells -- that part is correct. But if you subsequently press the "return" key on any child cell (the "return" key is a way to force Access Bridge Explorer to call getAccessibleContextInfo again and refresh the display), the cell name/description/etc will suddenly change to the name of the last rendered cell. If you repeat the same operation on all children of the table, they all end up having the same name/description/etc. This makes me think that the accessible context of JTable cells does not re-acquire the correct cell component each time -- and that is no fun...


Finally, as an aside, the reason all of this works this way is that JList/JTree and JTable in Swing use "virtual" nodes for their children, meaning the children nodes are instantiated on-demand only when they need to be rendered on screen. The reason it is implemented this way is that it allows Swing list/trees/table to scale to millions of children without having to have millions of child nodes instantiated at the same time in memory.

Suppose you have a list on screen with millions of elements but only 10 of them are visible. When Swing needs to paint the list on screen, it will render one child at a time (10 total) after asking the application to provide an object able to render a child given its index (the `getListItem` call below). So the code in JList to render the visible part of a list more or less looks like this:

```
for(var i = firstChildIndexToRender; i <= LastChildIndexToRender; i++) {
  // getListItem(i) returns a transient object containing data related to the i'th child.
  JListItem child = list.getListItem(i);
  child.renderOnScreen();
}
```

As you can see, the JList implementation does not require having millions of JListItem elements in memory. It only requires 1 at a time.

Now, even though the default implementation of `getListItem` is to acquire a string (the item text) from a list of strings stored in memory, which obviously does not scale to millions of elements, applications have the ability to provide a custom implementation of `getListItem`. So, an application could override `getListItem` to fetch strings from database for example, and there would be no need to have millions of items in memory at the same time.

So, the way an accessible context proxy for a list item works is similar to the way rendering works: it uses the `getListItem` method to re-acquire a child node when needed. The proxy essentially stores 2 things: the index of the child, and a pointer to the list. So, when the accessible context proxy is asked for the name/description/etc. of the child, it does something along the lines, in this case assuming it is being asking the "name" property:

```
// Re-acquire child
JListItem child = myList.getListItem(myChildIndex);
// If child supports accessibility, asks it for the accessible name
if (child instanceof Accessible) {
  return child.getAccessibleContext().getAccessibleName();
}
return null;
```

**Comment by jcsteh** [comment]
I totally understand why the information is fetched on-demand rather 
than fetching it all ahead of time. However, IMO, that still doesn't 
mean the contexts shouldn't be comparable. As you pointed out, the index 
is still the same, so there's no reason a) the proxy couldn't be cached 
if another instance of it currently exists; or b) that comparing two 
proxies for the same index couldn't return true. IMO, that's a major 
design flaw. Still, it's rather academic, since that's what we have to 
deal with and no one at Oracle seems to actually care about improving 
Java accessibility any more...

Your thought about checking the focused state is similar to what I was 
initially thinking, except I'd go a step further: if the focused state 
is present, check if NVDA's focus object still has the focused state, 
and if it does, just map the state change to the focus object and be 
done with it. That avoids a lot of extra calls. One flaw with this is 
that it won't allow us to map state changes to any other live object 
(e.g. if we start supporting state changes on the review object one 
day), but IMO, we just have to write that off as something we can't 
solve due to JAB deficiencies. Apart from that, does anyone see an issue 
with this?

**Comment by mohammad-suliman** [comment]
yes, this is a solution also, a tradeoff between performance and 
functionality.
what about caching  accessibleContextInfo on initialization of the jab 
context.
in that way, eq will never call the DLL, only accessible context infos 
will be compared. and a call to getAccessibleContextInfo when the object 
gets the focus will be saved.
the problem is: does that way will be reliable?
reliability could be improved as Renaud pointed out by adding more info 
to the equality test. however, more calls to the DLL are needed. caching 
here also could be used to improve efficiency.
what do you think?

**Comment by jcsteh** [comment]
The problem with caching on construction is that the context info can 
change at any time. Also, if you did cache it, it actually wouldn't be 
equal once the state change happens, so comparing wouldn't work.

**Comment by mohammad-suliman** [comment]
that's very correct, sorry, I missed that.

**Comment by Adriani90** [comment]
cc: @kaleev

**Comment by mwhapples** [comment]
I have been looking into this one. Probably the most informative documentation I can find is the Java documentation for AccessibleState https://docs.oracle.com/en/java/javase/17/docs/api/java.desktop/javax/accessibility/AccessibleState.html in particular for the states ACTIVE, TRANSIENT and MANAGES_DESCENDANTS.
From ACTIVE: In addition, this state is used to indicate the currently active child of a component such as a list, table, or tree. For example, the active child of a list is the child that is drawn with a rectangle around it.
From TRANSIENT: Indicates this object is transient. An assistive technology should not add a PropertyChange listener to an object with transient state, as that object will never generate any events. Transient objects are typically created to answer Java Accessibility method queries, but otherwise do not remain linked to the underlying object (for example, those objects underneath lists, tables, and trees in Swing, where only one actual UI Component does shared rendering duty for all of the data objects underneath the actual list/table/tree elements).
Interestingly I have never observed the TRANSIENT state through the access bridge using tools like Access Bridge Explorer. I don't know whether this is because TRANSIENT is not communicated by access bridge or whether TRANSIENT although in the API is never actually used.
Also this shows a difference in thinking, where in Java code you add listener instances to various objects. So you need never compare objects as you can add listeners only to the objects where you want to get events. Also you can register a different listener instance to each object, then within that listener you can be certain what the source is without checking it because only one object will trigger that listener instance.
I am still getting fully familiar with the Java Access Bridge API and so am still working out whether they expect a similar approach to be used for those using Java Access Bridge.

**Comment by mwhapples** [comment]
This feels a somewhat tough one to deal with. I have noticed a few things. With the Java demo application swingset2 in lists an tables the controls seem to have the transient state. For the tree view example I do not find the transient state. NOTE: This transient state is a Java accessibility state and is not mapped to anything in NVDA.
Another hint the object may be transient is the active state. This though is much more limited as its the focused item and potentially could throw a false positive as the top level window also can have the active state. The false positive can be avoided by checking if the control is a window.
So we cannot rely upon transient being there for all transient objects, but may be we could use it when it is present to know we will need to perform an alternative comparison technique. This alternative comparison may be slower to perform but should only occur for those objects which need it.
On the comparison, I would say it only needs the following:
1. The ancestor which manages it.
2. A list of child index values for each child level needed to get from the managing container to the control. For lists and tables this may only be a single index, but for a tree view it could easily be many more.
Whilst the solutions may seem imperfect they are still a step forward and I think we should go with the best we can do.

**Comment by Adriani90** [comment]
cc: @thgcode since you are familliar with JAB as well, maybe you have some suggestions here or can contribute to fixing this.
